File: samples/python/src/Graphs.py
from __future__ import annotations
from typing import TypeVar, Generic
from collections import deque
from queue import Queue

T = TypeVar("T")

class Graphs:

    @staticmethod
    def run_main():
        # nodes of a graph
        gnode0 = Graphs.GraphNode[str]("0")
        gnode1 = Graphs.GraphNode[str]("1")
        gnode2 = Graphs.GraphNode[str]("2")
        gnode3 = Graphs.GraphNode[str]("3")
        gnode4 = Graphs.GraphNode[str]("4")

        gnode0.add(gnode1, 10)
        gnode0.add(gnode2, 20)
        gnode2.add(gnode3, 15)
        gnode3.add(gnode2, 10)
        gnode3.add(gnode4, 25)
        gnode1.add(gnode4, 45)
        gnode4.add(gnode2, 55)

        # wrap to a graph
        graph: Graphs.Graph[str] = Graphs.Graph[str](gnode0)
        graphRoot: Graphs.GraphNode[str] = graph.get_root()
        graphRoot.add(Graphs.GraphNode[str]("5"), 11)
        gnode1.add(Graphs.GraphNode[str]("6"), 12)
        gnode2.add(Graphs.GraphNode[str]("7"), 13)

        Graphs.bfs(gnode0)

    @staticmethod
    def bfs(node: Graphs.GraphNode[str]):
        q: Queue[Graphs.GraphNode[str]] = Queue()
        q.put(node)
        v: set[Graphs.GraphNode[str]] = set[Graphs.GraphNode[str]]()
        steps = 0
        size = len(q.queue)
        while len(q.queue) > 0:
            x: Graphs.GraphNode[str] = q.get()
            size -= 1
            for child in x.children:
                if child in v:
                    continue
                q.put(child)
                v.add(child)

            if size == 0:
                steps += 1
                size = len(q.queue)

    class Graph(Generic[T]):
        def __init__(self, root: Graphs.GraphNode[T]):
            self.root: Graphs.GraphNode[T] = root

        def get_root(self) -> Graphs.GraphNode[T]:
            return self.root

    class GraphNode(Generic[T]):
        def get_value(self) -> T:
            return self.value

        def __init__(self, value: T):
            self.value = value
            self.children: list[Graphs.GraphNode[T]] = []
            self.childrenCost: list[Graphs.GraphNode[T]] = []

        def add(self, node: Graphs.GraphNode[T], cost: int):
            self.children.append(node)
            self.childrenCost.append(cost)

        def __str__(self):
            return str(self.value)

# custom extractor
class Extractor:
    # register the custom types
    @staticmethod
    def register_types() -> list[str]:
        return ["GraphNode"]

    @staticmethod
    def __str__(node: any):
        if type(node) == Graphs.GraphNode:
            return str(node.get_value())
        return ""

Graphs.run_main()

File: samples/ts/src/Graphs.ts
class Graphs {

    public static runMain() {
        new Graphs().test();
    }

    public test() {
        // nodes of a graph
        const gnode0: GraphNode<string> = new GraphNode<string>("0");
        const gnode1: GraphNode<string> = new GraphNode<string>("1");
        const gnode2: GraphNode<string> = new GraphNode<string>("2");
        const gnode3: GraphNode<string> = new GraphNode<string>("3");
        const gnode4: GraphNode<string> = new GraphNode<string>("4");
        gnode0.add(gnode1, 10);
        gnode0.add(gnode2, 20);
        gnode2.add(gnode3, 15);
        gnode3.add(gnode2, 10);
        gnode3.add(gnode4, 25);
        gnode1.add(gnode4, 45);
        gnode4.add(gnode2, 55);

        // // wrap to a graph
        const graph: Graph<string> = new Graph<string>(gnode0);
        const graphRoot: GraphNode<string> = graph.getRoot();
        graphRoot.add(new GraphNode<string>("5"), 11);
        gnode1.add(new GraphNode<string>("6"), 12);
        gnode2.add(new GraphNode<string>("7"), 13);

        Graphs.bfs(gnode0);
    }

    static bfs(node: GraphNode<string>) {
        const q: Array<GraphNode<string>> = new Array<GraphNode<string>>();
        q.push(node);
        const v: Set<GraphNode<string>> = new Set<GraphNode<string>>();
        let steps = 0;
        let size = q.length;
        while (q.length > 0) {
            const x: GraphNode<string> | undefined = q.shift();
            if (!x) {
                continue;
            }
            size--;
            for (const child of x.children) {
                if (v.has(child))
                    continue;
                q.push(child);
                v.add(child);
            }
            if (size == 0) {
                steps++;
                size = q.length;
                console.log("steps: ", steps);
            }
        }
    }
}

export class Graph<T> {

    private root: GraphNode<T>;

    public constructor(root: GraphNode<T>) {
        this.root = root;
    }

    public getRoot(): GraphNode<T> {
        return this.root;
    }
}

export class GraphNode<T> {

    private value: T;

    public getValue(): T {
        return this.value;
    }

    public children: Array<GraphNode<T>> = new Array<GraphNode<T>>();
    public childrenCost: Array<number> = new Array<number>();

    public constructor(value: T) {
        this.value = value;
    }

    public add(node: GraphNode<T>, cost: number) {
        this.children.push(node);
        this.childrenCost.push(cost);
    }

    public toString(): string {
        return String(this.value);
    }
}

// custom extractor
export class Extractor {
    // register the custom types
    public static registerTypes(): string[] {
        return ["GraphNode"];
    }

    public static toString(node: object): string {
        if (node instanceof GraphNode) {
            return (node as GraphNode<string>).getValue().toString();
        }
        return "";
    }
}
Graphs.runMain();